### ISSUE TITLE
New Upgrade V. 5.0.3 required

### ISSUE THREAD

**Issue by Osemudiahme** [sirosss/mg30-controller#2]
I would like to request for an update to version 5.0.3.
 I upgraded my nux MG-30 to the latest firmware version 5.0.3 but the last version you uploaded would not work with my MG-30 because of the firmware version 5.0.3 upgrade.

**Comment by Srijan619** [comment]
I would also be interested,  had a glance and not much should be different. I could contribute if needed as i have also standard understanding nux's MIDI implementation

**Comment by Osemudiahme** [comment]
> I would also be interested,  had a glance and not much should be different. I could contribute if needed as i have also standard understanding nux's MIDI implementation 

Your contribution will go a long way brother. I really don't have so much knowledge on these things as I just started learning DART programming language.

**Comment by sirosss** [comment]
> I would also be interested, had a glance and not much should be different. I could contribute if needed as i have also standard understanding nux's MIDI implementation

I don't have plan to update this project to support V5 firmware anytime soon. At least not to the level of functions of the current one.
The  most difficult part is not in programming, but in figuring out the preset structure. It is crucial for turning on/off effect blocks.

If you're interested, you can check the structure I worked out for V4 in the link below. MG-30 utilizes MIDI SysEx for communicating preset data between the unit and app (like QuickTone or this controller app), so it is all in binary format which is very cryptic. I guess V5 preset structure would be even more complex, since it allows each scene to have different effect types now.

[NUX MG-30 MIDI (V4)](https://docs.google.com/spreadsheets/d/1DgYG4zQBBtV4sSpklma7DKSFyL0kYqqRAHGBzRjhTmY/edit?usp=sharing)

**Comment by Srijan619** [comment]
Thank you for sharing this..this would be really helpful. I actually asked Nux distributor for more info regarding releasing details description like fractal does for their MIDI but he actually referred me to this repo and said this repo basically reveals most of the Nux' midi implementation. I will try to go through this documentation and see how feasible it is..

**Comment by Srijan619** [comment]
Providing an update.. the document was really helpful @sirosss . Thank you so much for that, even for V5 it is helpful. But there is huge bug in firmware V5 which I have encountered ane reported to Nux but they are saying they have not faced this issue....so in short the problem is with some effect change, it keeps sending same byte change info. Also found this issue with Tone editor in latest version, when I change some effect, it tries changing multiple stuff at once and they pointed out it could be driver issue or something but that's totally garbage..Tried downgrading to V4 and it works with no problem...

TLDR; Providing V5 support unless NUX fixes their shit is impossible, but anyway I am working on making a web version for firmware V4 of the editor and will soon make a beta release...stay tuned

**Comment by sirosss** [comment]
The preset structure of MG-30 was not properly designed to support these many features when it was released. Along the way, more features have been added to the unit, and NUX had to modify the preset structure accordingly. It is quite all over the place. We can't really blame them though, since these are quite drastic changes from the initial release.

Beware that V4 is not really bug free either. I found that presets made with V3 which can be transferred to V4 work on MG-30, but internally the preset data is not really correct (incorrect effect code, if I remember correctly).
I hope they have a better planning for their next unit. The unit sounds good and is affordable, but its MIDI and preset structure are pretty messy.

**Comment by Srijan619** [comment]
I agree, and V5 has basically put a nail in the coffin, just shows how messy it is. I spent few days and a back and forth email communication to Nux team but they were quite against opening these kind of info out .. Fractal has a proper wiki of their MIDI which has all APIs properly defined, the whole open sourcing could even help them design their unit better...but perhaps expecting too much..

**Comment by Srijan619** [comment]
https://srijan619.github.io/nuxOnline/

And its live for beta testing...try it out and make sure to use v4...alot of good stuff there..still needs few cosmetic improvements like knobs and etc but something to get started..
